crawl-ref/source/initfile.cc

```diff
@@ -5670,6 +5718,184 @@ bool parse_args(int argc, char **argv, bool rc_only)
+                    heading = nullptr;
+                    if (found_hover && new_hovered < 0)
+                        new_hovered = chosen_items.size();
+                    entry->hotkeys[0] = index_to_letter((i++)%52);
+                    chosen_items.emplace_back(entry);
+                }
+            }
+            if (!chosen_items.empty())
+                break;
+
+            show_type_response("No option names match.");
+        }
+
+        items = chosen_items;
+        string new_title = base_title;
+        if (!search_pat.empty())
+        {
+            string search = search_pat.tostring();
+            if (search.length() > 42)
+                search = search.substr(39)+"...";
+            search = replace_all(search, "<", "<<");
+            new_title += " <h>(Matches \""+search+"\")</h>";
+        }
+        set_title(new MenuEntry(new_title, MEL_TITLE));
+        reset();
+        update_menu(true);
+        if (new_hovered < 0)
+            set_hovered(all_items.size()-1);
+        else
+            set_hovered(new_hovered);
+    }
+
+    string get_keyhelp(bool) const override
+    {
+        return "<lightgrey>[<w>Up</w>|<w>Down</w>|<w>PgUp</w>|<w><<</w>"
+               "|<w>PgDn</w>|<w>></w>] select "
+               "[<w>Esc</w>] close "
+               "[<w>Ctrl-f</w>] find "
+               "[<w>Ctrl-s</w>] save "
+               "[<w>?</w>] help</lightgrey>";
+    }
+    int pre_process(int key) override
+    {
+        if (CONTROL('F') == key)
+            filter_items();
+        else if ('?' == key)
+            (static_cast<GameOption*>(items[last_hovered]->data))->show_help();
+        else if (CONTROL('S') == key)
+            _save_preferences();
+        else
+            return key;
+        return CK_NO_KEY;
+    }
+
+private:
+
+    text_pattern search_pat;
+    vector<MenuEntry*> all_items;
+    string base_title = "<w>Select a preference to set.</w>";
+};
+
+static string _option_line(const GameOption *option, int name_len, int text_len)
+{
+    auto name0 = option->name(), value0 = option->str();
+    if (static_cast<unsigned>(text_len) < value0.size())
+        value0.erase(0, value0.size()-text_len+3).insert(0, "...");
+    value0 = replace_all(value0, "<", "<<");
+    auto name = name0.c_str(), value = value0.c_str();
+    return make_stringf("%-*.*s%s", name_len, name_len, name, value);
+}
+
+// Show (and perhaps edit) options for the game.
+void edit_game_prefs()
+{
+    auto list = Options.get_sorted_options();
+    string selected;
+
+    // The caller should remove any user-provided formatting.
+    EGP_Menu menu(MF_SINGLESELECT | MF_ARROWS_SELECT
+                  | MF_ALLOW_FORMATTING | MF_INIT_HOVER);
+    menu.set_tag("option");
+
+    int i = 0;
+    for (const auto option : list)
+    {
+        EGP_MenuEntry *entry;
+        if (option->name().empty())
+            entry = new EGP_MenuEntry(option->str(), MEL_SUBTITLE);
+        else
+        {
```

Comment:
Same comment here re dedenting.

```suggestion
        if (option->name().empty())
        {
            entry = new EGP_MenuEntry(option->str(), MEL_SUBTITLE);
            continue;
        }
```

But honestly, it'd be much nicer to turn this whole `for` loop into a helper function that returns a `EGP_MenuEntry *entry`. I haven't dug into the code to see how feasible that is.

Reply:
That change wouldn't work as the loop doesn't end with the else block. It can go into a separate function, however, so I've done that.